File: Android-Client/app/src/test/java/com/grabtube/android/domain/usecase/ScanQRCodeUseCaseTest.kt
package com.grabtube.android.domain.usecase

import com.grabtube.android.domain.model.QRScanResult
import com.grabtube.android.domain.repository.QRScannerRepository
import io.mockk.coEvery
import io.mockk.coVerify
import io.mockk.mockk
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import java.time.LocalDateTime

class ScanQRCodeUseCaseTest {

    private lateinit var useCase: ScanQRCodeUseCase
    private lateinit var mockRepository: QRScannerRepository

    @Before
    fun setUp() {
        mockRepository = mockk()
        useCase = ScanQRCodeUseCase(mockRepository)
    }

    @Test
    fun `invoke returns success when repository scan succeeds`() = runTest {
        // Arrange
        val expectedResult = QRScanResult(
            rawValue = "https://youtube.com/watch?v=test",
            extractedUrl = "https://youtube.com/watch?v=test",
            scannedAt = LocalDateTime.now(),
            isValidUrl = true
        )
        coEvery { mockRepository.hasCameraPermission() } returns true
        coEvery { mockRepository.scanQRCode() } returns Result.success(expectedResult)

        // Act
        val result = useCase()

        // Assert
        assertTrue(result.isSuccess)
        assertEquals(expectedResult, result.getOrNull())
        coVerify { mockRepository.hasCameraPermission() }
        coVerify { mockRepository.scanQRCode() }
        coVerify(exactly = 0) { mockRepository.requestCameraPermission() }
    }

    @Test
    fun `invoke requests permission when not granted and succeeds`() = runTest {
        // Arrange
        val expectedResult = QRScanResult(
            rawValue = "https://youtube.com/watch?v=test",
            extractedUrl = "https://youtube.com/watch?v=test",
            scannedAt = LocalDateTime.now(),
            isValidUrl = true
        )
        coEvery { mockRepository.hasCameraPermission() } returns false
        coEvery { mockRepository.requestCameraPermission() } returns Result.success(Unit)
        coEvery { mockRepository.scanQRCode() } returns Result.success(expectedResult)

        // Act
        val result = useCase()

        // Assert
        assertTrue(result.isSuccess)
        assertEquals(expectedResult, result.getOrNull())
        coVerify { mockRepository.hasCameraPermission() }
        coVerify { mockRepository.requestCameraPermission() }
        coVerify { mockRepository.scanQRCode() }
    }

    @Test
    fun `invoke returns failure when permission denied`() = runTest {
        // Arrange
        coEvery { mockRepository.hasCameraPermission() } returns false
        coEvery { mockRepository.requestCameraPermission() } returns Result.failure(Exception("Permission denied"))

        // Act
        val result = useCase()

        // Assert
        assertTrue(result.isFailure)
        coVerify { mockRepository.hasCameraPermission() }
        coVerify { mockRepository.requestCameraPermission() }
        coVerify(exactly = 0) { mockRepository.scanQRCode() }
    }

    @Test
    fun `invoke returns failure when scan fails`() = runTest {
        // Arrange
        val expectedError = Exception("Scan failed")
        coEvery { mockRepository.hasCameraPermission() } returns true
        coEvery { mockRepository.scanQRCode() } returns Result.failure(expectedError)

        // Act
        val result = useCase()

        // Assert
        assertTrue(result.isFailure)
        assertEquals(expectedError, result.exceptionOrNull())
        coVerify { mockRepository.hasCameraPermission() }
        coVerify { mockRepository.scanQRCode() }
    }
}

File: Flutter-Client/test/unit/domain/usecases/scan_qr_code_usecase_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mocktail/mocktail.dart';
import 'package:dartz/dartz.dart';
import 'package:grabtube/domain/entities/qr_scan_result.dart';
import 'package:grabtube/domain/repositories/qr_scanner_repository.dart';
import 'package:grabtube/domain/usecases/scan_qr_code_usecase.dart';

class MockQRScannerRepository extends Mock implements QRScannerRepository {}

void main() {
  group('ScanQRCodeUseCase', () {
    late ScanQRCodeUseCase useCase;
    late MockQRScannerRepository mockRepository;

    setUp(() {
      mockRepository = MockQRScannerRepository();
      useCase = ScanQRCodeUseCase(mockRepository);
    });

    test('should return QR scan result when permission is granted', () async {
      // Arrange
      when(() => mockRepository.hasCameraPermission()).thenAnswer((_) async => true);
      when(() => mockRepository.scanQRCode()).thenAnswer(
        (_) async => Right(
          QRScanResult(
            rawValue: 'https://youtube.com/watch?v=test',
            extractedUrl: 'https://youtube.com/watch?v=test',
            scannedAt: DateTime(2023, 1, 1),
            isValidUrl: true,
          ),
        ),
      );

      // Act
      final result = await useCase();

      // Assert
      expect(result, isA<Right<String, QRScanResult>>());
      result.fold(
        (error) => fail('Expected success but got error: $error'),
        (scanResult) {
          expect(scanResult.rawValue, equals('https://youtube.com/watch?v=test'));
          expect(scanResult.extractedUrl, equals('https://youtube.com/watch?v=test'));
          expect(scanResult.isValidUrl, equals(true));
        },
      );
      verify(() => mockRepository.hasCameraPermission()).called(1);
      verify(() => mockRepository.scanQRCode()).called(1);
      verifyNever(() => mockRepository.requestCameraPermission());
    });

    test('should request permission and scan when permission is not granted', () async {
      // Arrange
      when(() => mockRepository.hasCameraPermission()).thenAnswer((_) async => false);
      when(() => mockRepository.requestCameraPermission()).thenAnswer((_) async => const Right(null));
      when(() => mockRepository.scanQRCode()).thenAnswer(
        (_) async => Right(
          QRScanResult(
            rawValue: 'https://example.com/video',
            extractedUrl: 'https://example.com/video',
            scannedAt: DateTime(2023, 1, 1),
            isValidUrl: true,
          ),
        ),
      );

      // Act
      final result = await useCase();

      // Assert
      expect(result, isA<Right<String, QRScanResult>>());
      result.fold(
        (error) => fail('Expected success but got error: $error'),
        (scanResult) {
          expect(scanResult.rawValue, equals('https://example.com/video'));
          expect(scanResult.extractedUrl, equals('https://example.com/video'));
          expect(scanResult.isValidUrl, equals(true));
        },
      );
      verify(() => mockRepository.hasCameraPermission()).called(1);
      verify(() => mockRepository.requestCameraPermission()).called(1);
      verify(() => mockRepository.scanQRCode()).called(1);
    });

    test('should return error when permission request fails', () async {
      // Arrange
      when(() => mockRepository.hasCameraPermission()).thenAnswer((_) async => false);
      when(() => mockRepository.requestCameraPermission()).thenAnswer((_) async => Left('Permission denied'));

      // Act
      final result = await useCase();

      // Assert
      expect(result, isA<Left<String, QRScanResult>>());
      result.fold(
        (error) => expect(error, equals('Permission denied')),
        (_) => fail('Expected error but got success'),
      );
      verify(() => mockRepository.hasCameraPermission()).called(1);
      verify(() => mockRepository.requestCameraPermission()).called(1);
      verifyNever(() => mockRepository.scanQRCode());
    });

    test('should return error when scanning fails', () async {
      // Arrange
      when(() => mockRepository.hasCameraPermission()).thenAnswer((_) async => true);
      when(() => mockRepository.scanQRCode()).thenAnswer((_) async => Left('Scanning failed'));

      // Act
      final result = await useCase();

      // Assert
      expect(result, isA<Left<String, QRScanResult>>());
      result.fold(
        (error) => expect(error, equals('Scanning failed')),
        (_) => fail('Expected error but got success'),
      );
      verify(() => mockRepository.hasCameraPermission()).called(1);
      verify(() => mockRepository.scanQRCode()).called(1);
      verifyNever(() => mockRepository.requestCameraPermission());
    });

    test('should return error when permission request returns null', () async {
      // Arrange
      when(() => mockRepository.hasCameraPermission()).thenAnswer((_) async => false);
      when(() => mockRepository.requestCameraPermission()).thenAnswer((_) async => Left('Permission denied'));

      // Act
      final result = await useCase();

      // Assert
      expect(result, isA<Left<String, QRScanResult>>());
      result.fold(
        (error) => expect(error, equals('Permission denied')),
        (_) => fail('Expected error but got success'),
      );
      verify(() => mockRepository.hasCameraPermission()).called(1);
      verify(() => mockRepository.requestCameraPermission()).called(1);
      verifyNever(() => mockRepository.scanQRCode());
    });
  });
}